### PR TITLE
adding devcontainer setup

### PR DESCRIPTION
**Pull request by zikalino**
Adding devcontainer setup, it requires Visual Studio Code Insiders Edition.

For more details please read note here:

https://zikalino.github.io/blog/2019/06/10/setting-azure-cli-development-environment-using-visual-studio-code-remote-development-with-containers/

When this PR is merged I could also move the note (or parts of it) to become official documentation.

When the project is opened from VSC Insiders:
- VSC will automatically build development container (after confirmation)
- Developers will be able to start debugging (set a breakpoint, etc.) immediately after the container is ready. No additional steps are required

This provides very consistent development environment.
I think this PR is ready to be merged, I would like to use it in my daily work and continue testing.
Perhaps some things will change, for instance we may discover that the environment could be setup in more convenient way.
We could also refer ready-made image instead of Dockerfile, if we decide to have any prebuilt public dev image for VSC.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by tjprescott** [comment]
@marstr for review since this is his wheelhouse.

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+RUN apt-get -y install software-properties-common
+RUN add-apt-repository ppa:deadsnakes/ppa
+RUN apt-get update
+RUN apt-get -y install python3.6 python3-pip python3-venv git
+
+# clone azure cli into /workspaces/azure-cli
+RUN mkdir /workspaces; cd /workspaces; git clone https://github.com/Azure/azure-cli.git
+
+# create environment in /env and set up 
+RUN python3 -m venv env
```
Comment (marstr) [inline .devcontainer/Dockerfile]: Given that we're running in a container, I think that the environment is already as isolated as it needs to be. Not using a virtual environment here may simplify life a little.

**Comment by marstr** [comment]
Not really to do with this PR directly, but it's a good place to start a team-wide conversation. :) IMHO, I don't like to check-in dev environment specific code. If it were me, I'd be deleting the solution files and `.vscode` folder that are currently in the repository. In my mind they're just one more set of things to keep in sync, especially when not everybody is using the same tools.

Now that said, given that this seems a little more involved to get setup than most configuration files.. Maybe it makes sense to add an `azdev` command to setup containerized debugging for VS Code users?

File: .devcontainer/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,23 @@
+
+# start from Ubuntu 16.04
+FROM ubuntu:xenial
```
[the remark this answers is not in the document]
Comment (zikalino) [reply .devcontainer/Dockerfile]: I can switch to bionic, np.
Actually I sometimes have some problems with installing packages in bionic in China, but always works if I use VPN.

**Comment by zikalino** [comment]
@marstr Yes, you are right, that maybe we should consider whethere these files should be merged or not. I actually created this PR to start the discussion :-)
Perhaps I could move it to azdev.
Maybe we should remove all vscode specific files from here.

Also, while talking about azure-cli-dev-tools, i found sln file which I tried, but had to update, as it was outdated, which means it's not probably used by anyone, and I am probably going to use VSC anyway, so perhaps instead of updating it i should just delete it.

File: .devcontainer/Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (marstr) [reply .devcontainer/Dockerfile]: Woah! That's really interesting :)

[the remark this answers is not in the document]
Comment (zikalino) [reply .devcontainer/Dockerfile]: anyway, I have one question regarding currently supported containers, but I thin I should create another thread for that.

**Comment by zikalino** [comment]
closing this one I will create another one with better solution

**Pull request closed without merging**